File: src/lib/oneEuro.ts
/**
 * 1€ filter (Casiez, Roussel & Vogel, CHI 2012).
 *
 * A plain low-pass filter forces a choice between jitter and lag. This one
 * adapts its cutoff to speed: it filters hard while the hand is nearly still
 * (killing tracker jitter) and opens up as the hand accelerates (so fast
 * strokes stay latency-free). That property is the reason air-drawing feels
 * attached to your fingertip instead of dragging behind it.
 */

function alphaFor(cutoffHz: number, dt: number): number {
  const tau = 1 / (2 * Math.PI * cutoffHz);
  return 1 / (1 + tau / dt);
}

class LowPass {
  private state: number | null = null;

  filter(value: number, alpha: number): number {
    this.state = this.state === null ? value : alpha * value + (1 - alpha) * this.state;
    return this.state;
  }

  get hasValue(): boolean {
    return this.state !== null;
  }

  reset(): void {
    this.state = null;
  }
}

export interface OneEuroConfig {
  /** Cutoff at zero speed, in Hz. Lower = steadier, but laggier. */
  minCutoff: number;
  /** How aggressively the cutoff opens with speed. Higher = less lag. */
  beta: number;
  /** Cutoff for the derivative estimate itself. */
  derivativeCutoff: number;
}

export class OneEuroFilter {
  private readonly value = new LowPass();
  private readonly derivative = new LowPass();
  private lastValue = 0;
  private lastTime: number | null = null;

  constructor(private config: OneEuroConfig) {}

  configure(config: Partial<OneEuroConfig>): void {
    this.config = { ...this.config, ...config };
  }

  /** @param time seconds (monotonic) */
  filter(value: number, time: number): number {
    if (this.lastTime === null) {
      this.lastTime = time;
      this.lastValue = value;
      return this.value.filter(value, 1);
    }

    // Guard against duplicate timestamps and against a stalled tab producing a
    // huge dt, which would otherwise make alpha ~1 and pass jitter straight through.
    const dt = Math.min(Math.max(time - this.lastTime, 1 / 240), 1 / 5);
    this.lastTime = time;

    const speed = (value - this.lastValue) / dt;
    this.lastValue = value;

    const smoothedSpeed = this.derivative.filter(
      speed,
      alphaFor(this.config.derivativeCutoff, dt),
    );

    const cutoff = this.config.minCutoff + this.config.beta * Math.abs(smoothedSpeed);
    return this.value.filter(value, alphaFor(cutoff, dt));
  }

  reset(): void {
    this.value.reset();
    this.derivative.reset();
    this.lastTime = null;
    this.lastValue = 0;
  }
}

/** Two independent 1€ filters, for a 2D point. */
export class OneEuroPoint {
  private readonly fx: OneEuroFilter;
  private readonly fy: OneEuroFilter;

  constructor(config: OneEuroConfig) {
    this.fx = new OneEuroFilter(config);
    this.fy = new OneEuroFilter(config);
  }

  configure(config: Partial<OneEuroConfig>): void {
    this.fx.configure(config);
    this.fy.configure(config);
  }

  filter(x: number, y: number, time: number): { x: number; y: number } {
    return { x: this.fx.filter(x, time), y: this.fy.filter(y, time) };
  }

  reset(): void {
    this.fx.reset();
    this.fy.reset();
  }
}

File: drawone/lib/one_euro.py
"""1€ filter (Casiez, Roussel & Vogel, CHI 2012).

A plain low-pass filter forces a choice between jitter and lag. This one adapts
its cutoff to speed: it filters hard while the hand is nearly still (killing
tracker jitter) and opens up as the hand accelerates (so fast strokes stay
latency-free). That property is the reason air-drawing feels attached to your
fingertip instead of dragging behind it.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

from ..types import Vec2


def _alpha_for(cutoff_hz: float, dt: float) -> float:
    tau = 1.0 / (2.0 * math.pi * cutoff_hz)
    return 1.0 / (1.0 + tau / dt)


class _LowPass:
    def __init__(self) -> None:
        self._state: Optional[float] = None

    def filter(self, value: float, alpha: float) -> float:
        self._state = (
            value if self._state is None else alpha * value + (1 - alpha) * self._state
        )
        return self._state

    @property
    def has_value(self) -> bool:
        return self._state is not None

    def reset(self) -> None:
        self._state = None


@dataclass(frozen=True)
class OneEuroConfig:
    #: Cutoff at zero speed, in Hz. Lower = steadier, but laggier.
    min_cutoff: float
    #: How aggressively the cutoff opens with speed. Higher = less lag.
    beta: float
    #: Cutoff for the derivative estimate itself.
    derivative_cutoff: float = 1.0


class OneEuroFilter:
    def __init__(self, config: OneEuroConfig) -> None:
        self._config = config
        self._value = _LowPass()
        self._derivative = _LowPass()
        self._last_value = 0.0
        self._last_time: Optional[float] = None

    def configure(self, config: OneEuroConfig) -> None:
        self._config = config

    def filter(self, value: float, time: float) -> float:
        """`time` is in seconds, monotonic."""
        if self._last_time is None:
            self._last_time = time
            self._last_value = value
            return self._value.filter(value, 1.0)

        # Guard against duplicate timestamps and against a stalled process
        # producing a huge dt, which would otherwise make alpha ~1 and pass
        # jitter straight through.
        dt = min(max(time - self._last_time, 1.0 / 240.0), 1.0 / 5.0)
        self._last_time = time

        speed = (value - self._last_value) / dt
        self._last_value = value

        smoothed_speed = self._derivative.filter(
            speed, _alpha_for(self._config.derivative_cutoff, dt)
        )

        cutoff = self._config.min_cutoff + self._config.beta * abs(smoothed_speed)
        return self._value.filter(value, _alpha_for(cutoff, dt))

    def reset(self) -> None:
        self._value.reset()
        self._derivative.reset()
        self._last_time = None
        self._last_value = 0.0


class OneEuroPoint:
    """Two independent 1€ filters, for a 2D point."""

    def __init__(self, config: OneEuroConfig) -> None:
        self._fx = OneEuroFilter(config)
        self._fy = OneEuroFilter(config)

    def configure(self, config: OneEuroConfig) -> None:
        self._fx.configure(config)
        self._fy.configure(config)

    def filter(self, x: float, y: float, time: float) -> Vec2:
        return Vec2(self._fx.filter(x, time), self._fy.filter(y, time))

    def reset(self) -> None:
        self._fx.reset()
        self._fy.reset()
